solidity/contracts/contracts/QuantumGravityBridge.sol

```diff
@@ -0,0 +1,314 @@
+        bytes32 _digest,
+        Signature calldata _sig
+    ) private pure returns (bool) {
+        bytes32 digest_eip191 = ECDSA.toEthSignedMessageHash(_digest);
+
+        return _signer == ECDSA.recover(digest_eip191, _sig.v, _sig.r, _sig.s);
+    }
+
+    /// @dev Make a domain-separated commitment to the validator set.
+    /// A hash of all relevant information about the validator set.
+    /// The format of the hash is:
+    ///     keccak256(bridge_id, bytes32("checkpoint"), nonce, power_threshold, validator_set_hash)
+    /// The elements in the validator set should be monotonically decreasing by power.
+    /// @param _bridge_id Bridge ID.
+    /// @param _nonce Nonce.
+    /// @param _powerThreshold The voting power threshold.
+    /// @param _validatorSetHash Validator set hash.
+    function domainSeparateValidatorSetHash(
+        bytes32 _bridge_id,
+        uint256 _nonce,
+        uint256 _powerThreshold,
+        bytes32 _validatorSetHash
+    ) private pure returns (bytes32) {
+        // bytes32 encoding of the string "checkpoint"
+        bytes32 methodName = 0x636865636b706f696e7400000000000000000000000000000000000000000000;
+
+        bytes32 c = keccak256(abi.encode(_bridge_id, methodName, _nonce, _powerThreshold, _validatorSetHash));
+
+        return c;
+    }
+
+    /// @dev Make a domain-separated commitment to a message root.
+    /// A hash of all relevant information about a message root.
+    /// The format of the hash is:
+    ///     keccak256(bridge_id, bytes32("transactionBatch"), nonce, message_root)
+    /// @param _bridge_id Bridge ID.
+    /// @param _nonce Nonce.
+    /// @param _messageRoot Message root.
+    function domainSeparateMessageRoot(
+        bytes32 _bridge_id,
+        uint256 _nonce,
+        bytes32 _messageRoot
+    ) private pure returns (bytes32) {
+        // bytes32 encoding of the string "transactionBatch"
+        bytes32 methodName = 0x7472616e73616374696f6e426174636800000000000000000000000000000000;
+
+        bytes32 c = keccak256(abi.encode(_bridge_id, methodName, _nonce, _messageRoot));
+
+        return c;
+    }
+
+    /// @dev Checks that enough voting power signed over a digest.
+    /// @param _currentValidators The current validators.
+    /// @param _sigs The current validators' signatures.
+    /// @param _digest This is what we are checking they have signed.
+    /// @param _powerThreshold At least this much power must have signed.
+    function checkValidatorSignatures(
+        // The current validator set and their powers
+        Validator[] calldata _currentValidators,
+        Signature[] calldata _sigs,
+        bytes32 _digest,
+        uint256 _powerThreshold
+    ) private pure {
+        uint256 cumulativePower = 0;
+
+        for (uint256 i = 0; i < _currentValidators.length; i++) {
+            // Check that the current validator has signed off on the hash.
+            if (!verifySig(_currentValidators[i].addr, _digest, _sigs[i])) {
```

Comment:
We do not need to check for duplicates here because the assumption is that +2/3 of validators are honest and would never sign an update with duplicates?

Reply:
Right